### ISSUE TITLE
rsyslog imfile not sending logs after reboot

### ISSUE THREAD

**Issue by Teressss** [rsyslog/rsyslog-pkg-ubuntu#80]
I tried newest 8.35 but it gives me tons of omfile errors. So I reverted back to 8.16 where it runs OK. But after reboot on my ElasticSearch server I am recieving only system and kernel logs, not the logs from imfile. I have to manually restart resyslog to get those logs back.

My conf:
```*.* @172.16.1.34:10514;json-template
module(load="imfile" PollingInterval="10")
input(type="imfile"
      File="/var/log/miner/ethminer/ethminer.log"
      Tag="ethminer"
      Severity="info"
      Facility="local7")

**Comment by davidelang** [comment]
There was a change in the format of the state file in that window. delete the 
state files and you should be good with the older version.

what were the omfile problems you ran into?

David Lang

**Comment by Teressss** [comment]
David I have defined no StateFile for this log, so how can I delete it?

**Comment by Teressss** [comment]
On 8.35 I had these errors about 20 times a minute:
`Jun 07 18:50:55 KACIK-rig rsyslogd:  warning |  action 'action 2' suspended (module 'builtin:omfile'), retry 0. There should be messages before this one giving the reason for suspension. [v8.35.0 try http://www.rsyslog.com/e/2007 ]
Jun 07 18:50:55 KACIK-rig rsyslogd:  warning |  action 'action 2' suspended (module 'builtin:omfile'), retry 0. There should be messages before this one giving the reason for suspension. [v8.35.0 try http://www.rsyslog.com/e/2007 ]`

**Comment by Teressss** [comment]
And because of GDPR all Rsyslog pages related to these errors are not working, so advice "try http://www.rsyslog.com/e/2007" is unusable...

**Comment by davidelang** [comment]
On Thu, 7 Jun 2018, Teressss wrote:

> And because of GDPR all Rsyslog pages related to these errors are not working, so advice "try http://www.rsyslog.com/e/2007" is unusable...

these links have not been very usefor for quite a while

**Comment by davidelang** [comment]
can you open another issue on this and post your full config? omfile is very 
fast, so unless something odd is configured, it almost never suspends.

**Comment by davidelang** [comment]
check in your workdirectory, I think there is one made by default for every 
imfile input.

**Comment by Teressss** [comment]
Open another issue? I just opened this one, I don't understand. I wrote already I deeted all statefiles one post ago.

**Comment by Teressss** [comment]
This is my whole config:
`:programname, contains, "CRON" ~
#:msg, contains, "rsyslog" ~
#:programname, contains, "systemd" ~


set $!client = 'MineIt';

$template logdna,"<%PRI%>%protocol-version% %timestamp:::date-rfc3339% %HOSTNAME% %app-name% %procid% %msgid% [logdna@48350 key=\"49bebdc176fcfa7a4f8e4031b6884a\"] %msg%\n"

template(name="json-template"
  type="list") {
    constant(value="{")
      constant(value="\"@timestamp\":\"")     property(name="timereported" dateFormat="rfc3339")
      constant(value="\",\"@version\":\"1")
      constant(value="\",\"message\":\"")     property(name="msg" format="json")
      constant(value="\",\"host\":\"")  property(name="hostname")
      constant(value="\",\"log_level\":\"")    property(name="syslogseverity-text")
      constant(value="\",\"app\":\"") property(name="programname")
      constant(value="\",\"procid\":\"")      property(name="procid")
      constant(value="\",\"client\":\"")  property(name="$!client" format="json")
    constant(value="\"}\n")
}

#--------------------- SYSTEM ---------------------
#module(load="imklog" permitnonkernelfacility="on")
#module(load="imuxsock")
*.* @213.220.247.208:10514;json-template
*.* @@syslog-a.logdna.com:514;logdna



module(load="imfile" PollingInterval="10")
# ---------------------
#       CLAYMORE
# ---------------------
input(type="imfile"
      File="/var/log/miner/claymore/lastrun_noappend.log"
      Tag="claymore"
      Severity="info"
      Facility="local7")`

**Comment by davidelang** [comment]
I was meaning an issue for the omfile problem. as for the statefiles, our 
messages crossed.

**Comment by Teressss** [comment]
My topic of this problem is exactly what it describes my problem. So I really have no idea whu am I to open a new ticket and write there the same.

**Comment by davidelang** [comment]
I am not seeing any action here that writes to disk, so I don't see anywhere 
that omfile would be invoked.

there are also no inputs defined (other than imfile) so how can you be getting 
other logs?

**Comment by Teressss** [comment]
Yes exactly. There si nowhere any action that writes to disk. But still getting this error. So what am I to do now?...

**Comment by davidelang** [comment]
On Thu, 7 Jun 2018, Teressss wrote:

> My topic of this problem is exactly what it describes my problem. So I really 
> have no idea whu am I to open a new ticket and write there the same.

you said you had two problems

1. when you went to 8.35 you were getting a flood of omfile errors

2. when you moved back down to 8.16 the imfile is no longer working

This topic is about the imfile problem, I was suggesting a different topic to 
deal with the omfile problem.

but we can work on both here if you want.

**Comment by Teressss** [comment]
But I told that I am back to 8.16 so I am dealnig only with problems on 8.16 imfile problem.

**Comment by davidelang** [comment]
On Thu, 7 Jun 2018, Teressss wrote:

> Yes exactly. There si nowhere any action that writes to disk. But still getting this error. So what am I to do now?...

In the config you posted, there is also nothing that accepts logs from anything 
other than imfile, but you post that all the other logs work...

**Comment by davidelang** [comment]
On Thu, 7 Jun 2018, Teressss wrote:

> Yes exactly. There si nowhere any action that writes to disk. But still getting this error. So what am I to do now?...

it says it's happening for action 2, so we need to look at the file and the 
second action there is going to be something that invokes omfile.

**Comment by Teressss** [comment]
"In the config you posted, there is also nothing that accepts logs from anything
other than imfile, but you post that all the other logs work..."
in rsyslog.conf by default is:
module(load="imuxsock") # provides support for local system logging
module(load="imklog")   # provides kernel logging support
I have all rsyslog files at default. Only added my own posted above conf to rsyslog.d

**Comment by davidelang** [comment]
On Thu, 7 Jun 2018, Teressss wrote:

> But I told that I am back to 8.16 so I am dealnig only with problems on 8.16 imfile problem.

Ok, I will stop trying to figure out what went wrong on 8.35

but we still don't have the full config based on your statements about what's 
happening

**Comment by Teressss** [comment]
OK I will post rsyslog.conf and 50-default, even they are in defaults...

**Comment by Teressss** [comment]
`root@hive5-3-3:/var/spool/rsyslog# cat /etc/rsyslog.conf
#  /etc/rsyslog.conf	Configuration file for rsyslog.
#
#			For more information see
#			/usr/share/doc/rsyslog-doc/html/rsyslog_conf.html
#
#  Default logging rules can be found in /etc/rsyslog.d/50-default.conf


#################
#### MODULES ####
#################

module(load="imuxsock") # provides support for local system logging
module(load="imklog")   # provides kernel logging support
#module(load="immark")  # provides --MARK-- message capability

# provides UDP syslog reception
#module(load="imudp")
#input(type="imudp" port="514")

# provides TCP syslog reception
#module(load="imtcp")
#input(type="imtcp" port="514")

# Enable non-kernel facility klog messages
$KLogPermitNonKernelFacility on

###########################
#### GLOBAL DIRECTIVES ####
###########################

#
# Use traditional timestamp format.
# To enable high precision timestamps, comment out the following line.
#
$ActionFileDefaultTemplate RSYSLOG_TraditionalFileFormat

# Filter duplicated messages
$RepeatedMsgReduction on

#
# Set the default permissions for all log files.
#
$FileOwner syslog
$FileGroup adm
$FileCreateMode 0640
$DirCreateMode 0755
$Umask 0022
$PrivDropToUser syslog
$PrivDropToGroup syslog

#
# Where to place spool and state files
#
$WorkDirectory /var/spool/rsyslog

#
# Include all config files in /etc/rsyslog.d/
#
$IncludeConfig /etc/rsyslog.d/*.conf`

**Comment by davidelang** [comment]
On Thu, 7 Jun 2018, Teressss wrote:

> "In the config you posted, there is also nothing that accepts logs from anything
> other than imfile, but you post that all the other logs work..."
> in rsyslog.conf by default is:
> module(load="imuxsock") # provides support for local system logging
> module(load="imklog")   # provides kernel logging support
> I have all rsyslog files at default. Only added my own posted above conf to rsyslog.d

I don't know what your 'default files' are

in the file you posted, the lines you quote are not there.

I would need to see the full config (rsyslog.conf and any files that are 
included)

**Comment by Teressss** [comment]
`root@hive5-3-3:/var/spool/rsyslog# cat /etc/rsyslog.d/50-default.conf
#  Default rules for rsyslog.
#
#			For more information see rsyslog.conf(5) and /etc/rsyslog.conf

#
# First some standard log files.  Log by facility.
#
auth,authpriv.*			/var/log/auth.log
*.*;auth,authpriv.none		-/var/log/syslog
#cron.*				/var/log/cron.log
#daemon.*			-/var/log/daemon.log
kern.*				-/var/log/kern.log
#lpr.*				-/var/log/lpr.log
mail.*				-/var/log/mail.log
#user.*				-/var/log/user.log

#
# Logging for the mail system.  Split it up so that
# it is easy to write scripts to parse these files.
#
#mail.info			-/var/log/mail.info
#mail.warn			-/var/log/mail.warn
mail.err			/var/log/mail.err

#
# Logging for INN news system.
#
news.crit			/var/log/news/news.crit
news.err			/var/log/news/news.err
news.notice			-/var/log/news/news.notice

#
# Some "catch-all" log files.
#
#*.=debug;\
#	auth,authpriv.none;\
#	news.none;mail.none	-/var/log/debug
#*.=info;*.=notice;*.=warn;\
#	auth,authpriv.none;\
#	cron,daemon.none;\
#	mail,news.none		-/var/log/messages

#
# Emergencies are sent to everybody logged in.
#
*.emerg                                :omusrmsg:*

#
# I like to have messages displayed on the console, but only on a virtual
# console I usually leave idle.
#
#daemon,mail.*;\
#	news.=crit;news.=err;news.=notice;\
#	*.=debug;*.=info;\
#	*.=notice;*.=warn	/dev/tty8

# The named pipe /dev/xconsole is for the `xconsole' utility.  To use it,
# you must invoke `xconsole' with the `-file' option:
#
#    $ xconsole -file /dev/xconsole [...]
#
# NOTE: adjust the list below, or you'll go crazy if you have a reasonably
#      busy site..
#
daemon.*;mail.*;\
	news.err;\
	*.=debug;*.=info;\
	*.=notice;*.=warn	|/dev/xconsole`

**Comment by davidelang** [comment]
is your custom config file loaded before or after the 50-* included file?

David Lang

P.S. I've got to run to a meeting, so further replys will be delayed

**Comment by Teressss** [comment]
I have no idea when is my.conf loaded. It is located in rsyslog.d/my.conf

**Comment by davidelang** [comment]
it's loaded in alphabetical order, so it would be after the other config.

David Lang

**Comment by Teressss** [comment]
> Thanks for info about the order, but I think it's not causing this problem.

**Comment by davidelang** [comment]
On Thu, 7 Jun 2018, Teressss wrote:

>> Thanks for info about the order, but I think it's not causing this problem.

I needed to figure out the order to understand what the config is that rsyslog 
is seeing.

when there is an include, to rsyslog, it's as if the text from those files was 
pasted in at that point in the config file. Since we are looking for action 2, 
we needed to know which file to look at.

**Comment by Teressss** [comment]
You are mixing two problems now. Actions problems were on 8.32. On 8.16 I have problem, that all logs are running ok after reboot, only the log from the imgile in not gettig to my Elastic. I have to restart rsyslog and till next reboot it is ok.

**Comment by Teressss** [comment]
Do you have any idea why I am getting on my logserver all logs, only logs from that imfile are missing after reboot?

**Comment by davidelang** [comment]
On Fri, 8 Jun 2018, Teressss wrote:

> Do you have any idea why I am getting on my logserver all logs, only logs from that imfile are missing after reboot?

my idea is that the new version of rsyslog created statefiles that the old 
version can't understand. But you say that you went to the workdir and deleted 
the statefiles, so I don't know what else it could be (unless there are 
statefiles in the directory of the files being read or something like that)

unfortuanntly, ith 8.16 being so old (2+ years), it's not something that 
the community is going to spend a lot of time working on (hich is part of why I 
was trying to figure out why8.35 gave you problems so that you could switch back 
to it)

**Comment by Teressss** [comment]
All logs I am sending, system and kernel logs and the miner log are in /var/log..... and /var/log is mounted in RAM, it is recreated every time machine boots, so there cannnot be any leftover from the past. 

8.35 gave me tons of that action omfile errors, really tons. I mean 5 lines of y logs and then 20 errror lines and again and again. I check 10 times rsyslog.conf, 50-default.. and my conf, nowhere is anything about omf. I really have no idea where it could come from. Googled a lot and saw lot of people have problems with it.

**Comment by davidelang** [comment]
On Fri, 8 Jun 2018, Teressss wrote:

> All logs I am sending, system and kernel logs and the miner log are in 
> /var/log..... and /var/log is mounted in RAM, it is recreated every time 
> machine boots, so there cannnot be any leftover from the past.

the configs you posted include:

$WorkDirectory /var/spool/rsyslog

so there can be stuff left there (that's the first place I would look for the 
state files, the other would be where imfile is reading from

> 8.35 gave me tons of that action omfile errors, really tons. I mean 5 lines of 
> y logs and then 20 errror lines and again and again. I check 10 times 
> rsyslog.conf, 50-default.. and my conf, nowhere is anything about omf. I 
> really have no idea where it could come from.

omfile is the module used to write logs to files on disk, so just about 
everything in the /etc/rsyslog.d/50-* file is using omfile (action 2 appears to 
be writing to /var/log/syslog if I'm reading things correctly)

any time when you have

filter /path/to/file

you are using the omfile module.

> Googled a lot and saw lot of people have problems with it.

This is not a problem that's been reported to us as far as I can remember.

We may end up needing to get debug logs for a few seconds while this is 
happening to see what's going on.

David Lang

**Comment by Teressss** [comment]
"$WorkDirectory /var/spool/rsyslog"
I told you already that I deleted * in it dozens times - makes no difference.

"We may end up needing to get debug logs for a few seconds"
where can I find those logs you need?

**Comment by davidelang** [comment]
On Fri, 8 Jun 2018, Teressss wrote:

> "$WorkDirectory /var/spool/rsyslog"
> I told you already that I deleted * in it dozens times - makes no difference.

I understand, I was just pointing out that your statement "nothing could 
persist because /var/log is a ramdisk" wasn't the whole picture.

> "We may end up needing to get debug logs for a few seconds"
> where can I find those logs you need?


if you can start rsyslog with -dn and run it just long enough to get the first 
batch of the bad messages, that output will give us everything we need and more.

The more complicated way (that will produce less massive output) would be to 
setup debug on demand and then turn it on for a few seconds once the problem 
logs are being generated.

one way to do this is documented at 
https://www.rsyslog.com/how-to-use-debug-on-demand/

there are also ways to set the options in the global() section of the config 
rather than as environment variables before starting rsyslog.

David Lang

**Comment by Teressss** [comment]
I have no idea how to start rsyslog with -dn. I can start via service rsyslog start ot via systemctl

**Comment by davidelang** [comment]
it's similar to the debug on demand page

first top rsyslog via service or systemctl

then do rsyslogd -dn >/path/to/file

David Lang

**Comment by Teressss** [comment]
I've found that when I use old syntax for setitng up a file.log forwarding then everything works. Only with new sysntax is problem after reboot. Maybe it helps somebody.

Also your rsyslogd -dn >/path/to/file gives me some errors, but still that conf is working great with old imfile syntax.

Also I have no problem exit that debug, I even cannot edit my conf file. :(

**Comment by Teressss** [comment]
Oh God, kill -USR1 `cat /var/run/rsyslogd.pid` is not working, now I have totally unusable system :(

**Comment by davidelang** [comment]
you can always do

ps ax |grep rsyslog

to find what pid rsyslog has

doing a control-c on the terminal where you ran rsyslog -dn should stop it as 
well

David Lang

**Comment by Teressss** [comment]
I killed all rsyslog processes, rebooted, no effect. When I try edit my conf I get always this:
```1524.114799822:main thread    : rsyslogd 8.16.0 startup, module path '', cwd:/var/spool/rsyslog
1524.114996789:main thread    : caller requested object 'net', not found (iRet -3003)
1524.115015299:main thread    : Requested to load module 'lmnet'
1524.115027100:main thread    : loading module '/usr/lib/rsyslog/lmnet.so'
1524.115202318:main thread    : module lmnet of type 2 being loaded (keepType=0).
1524.115211980:main thread    : entry point 'isCompatibleWithFeature' not present in module
1524.115217831:main thread    : entry point 'setModCnf' not present in module
1524.115223612:main thread    : entry point 'getModCnfName' not present in module
and tons of other lines

I have to completelly wipe whole system a confgure from scratch because of this stupid rsyslog. Oh God.

**Comment by davidelang** [comment]
what changes did you make to the config file?

after a reboot, the fact that you stopped rsyslog and ran it with the -dn switch 
is not going to matter.

the fact that it's looking for lmnet.so indcates that there is a module being 
loaded that isn't on your system.

modules only get loaded if the config files say to do so.

The other possible thing is that you may have two versions of rsyslog installed 
on your system (since you did go to a new version and then revert, it may not 
have removed all the files from one of the versions)

try reinstalling whichever version you are trying to use.

David Lang

**Comment by Teressss** [comment]
I did changed nothing in my config. Only ran your command for debug mode and now I'm totally stucked.

**Comment by davidelang** [comment]
running  rsyslog -dn would have no effect after a reboot. you mentioned that you 
got things orking more by switching to the old syntax or something like that.

try removing your my.conf file (or whatever it's named) so that the system only 
has the default config.

I don't know what distro you are running, but purging rsyslog and reinstalling 
may help if it's confused between files from different versions being there.

David Lang

**Comment by Teressss** [comment]
Solved. Problem was your command ovewritten my config. Redoit and it is running ok. I ignore those errors, when my logs are correctly shipping. 

Only wonder why old syntax works, but new syntax is not working.

**Comment by Teressss** [comment]
You wrote me "rsyslogd -dn >/path/to/file" so I thought path to my config like Filebeat debug have and that kill my my conf. So if you would write "rsyslogd -dn >/path/to/debugfileoutput" then I would not run into that problem:)

**Comment by davidelang** [comment]
On Sun, 10 Jun 2018, Teressss wrote:

> You wrote me "rsyslogd -dn >/path/to/file" so I thought path to my config like 
> Filebeat debug have and that kill my my conf. So if you would write "rsyslogd 
> -dn >/path/to/debugfileoutput" then I would not run into that problem:)

ok, sorry about that. I made an incorrect assumption that you would understand 
that output was implied by the >

so are you up and running now? or do you still need help with something?

**Comment by Teressss** [comment]
Yes, it's is running OK on old syntax. Errors shown by debug I ignore. I only have fear when I on some other systems one day would need to switch to newer version of Ubuntu I would be facing the problem with omfile errors we discussed earlier. 
But anyway big thanks for your time and effort.
